include/util/binary_heap.hpp

```diff
@@ -16,6 +16,40 @@ namespace osrm
 namespace util
 {
 
+template <typename NodeID, typename Key> class GenerationArrayStorage
+{
+    using GenerationCounter = std::uint16_t;
+  public:
+    explicit GenerationArrayStorage(std::size_t size) : positions(size, 0), generation(0), generations(size, 0) {}
+
+    Key &operator[](NodeID node) {
+        generation[node] = generation;
+        return positions[node];
+    }
+
+    Key peek_index(const NodeID node) const {
+        if (generations[node] < generation)
```

Comment:
If the initial generation is 0 and `Clear` was not called before then  `generations[node] < generation` will be false, so `peak_index` will return 0 for nodes that were not touched. It is different wrt `UnorderedMapStorage` and different in other way wrt `ArrayStorage` 

Also i guess  if `ArrayStorage` will be removed then `heap[0].weight = std::numeric_limits<Weight>::min();` can be also removed from `BinaryHeap`